ros2cli/ros2cli/node/strategy.py

```diff
@@ -25,21 +25,47 @@ class NodeStrategy:
     def __init__(self, args):
         use_daemon = not getattr(args, 'no_daemon', False)
         if use_daemon and is_daemon_running(args):
-            self.node = DaemonNode(args)
+            self._daemon_node = DaemonNode(args)
```

Comment:
This looks like a race between `is_daemon_running` and creating the `DaemonNode`. You can end up with a not-working daemon node (and access of its `None` proxy would raise) but no fallback.

Reply:
> This looks like a race between is_daemon_running and creating the DaemonNode

There is a race, but I don't see a way around it.

> You can end up with a not-working daemon node (and access of its None proxy would raise) but no fallback.

That is true for the current [`NodeStrategy` implementation](https://github.com/ros2/ros2cli/blob/b5b53263410a594d6211cc2a7b6ee827f966b1aa/ros2cli/ros2cli/node/strategy.py#L28). With this patch, [this condition](https://github.com/ros2/ros2cli/pull/499/files#diff-8a54fb07e6acc0ff6c8e4d8c731dca06R59) would evaluate to `False` in that case and a `DirectNode` would get instantiated instead.

Reply:
But that fallback is opaque to the caller and the function might return incomplete information since the direct node didn't have any time to discover, no?

Reply:
Right. But that's expected. Same goes if the daemon wasn't running at the moment and had to be spawned, in which case there's no real benefit in using the daemon (which isn't a silver bullet to cope with discovery latency anyways, we simply _hope_ the daemon and the rest of the ROS ecosystem have been around long enough for that not to be a problem).